control-plane/api/v1alpha1/serviceresolver_types.go

```diff
@@ -374,9 +446,76 @@ func (in *ServiceResolverRedirect) toConsul() *capi.ServiceResolverRedirect {
 		Datacenter:    in.Datacenter,
 		Partition:     in.Partition,
 		Peer:          in.Peer,
+		SamenessGroup: in.SamenessGroup,
 	}
 }
 
+func (in *ServiceResolverRedirect) validate(path *field.Path, consulMeta common.ConsulMeta) field.ErrorList {
+	var errs field.ErrorList
+	if in == nil {
+		return nil
+	}
+
+	asJSON, _ := json.Marshal(in)
+	if in.isEmpty() {
+		errs = append(errs, field.Invalid(path, string(asJSON),
+			"service resolver redirect cannot be empty"))
+	}
+
+	if consulMeta.Partition != "default" && in.Datacenter != "" {
+		errs = append(errs, field.Invalid(path.Child("datacenter"), in.Datacenter,
+			"cross-datacenter redirect is only supported in the default partition"))
+	}
+	if consulMeta.Partition != in.Partition && in.Datacenter != "" {
+		errs = append(errs, field.Invalid(path.Child("partition"), in.Partition,
+			"cross-datacenter and cross-partition redirect is not supported"))
+	}
+
+	switch {
+	case in.SamenessGroup != "" && in.ServiceSubset != "":
+		errs = append(errs, field.Invalid(path, string(asJSON),
+			"redirect.samenessGroup cannot be set with redirect.serviceSubset"))
+	case in.SamenessGroup != "" && in.Partition != "":
+		errs = append(errs, field.Invalid(path, string(asJSON),
+			"redirect.partition cannot be set with redirect.samenessGroup"))
```

Comment:
I would just check again, but I think `redirect` might already be covered by the path?